README.md
# X-ray photoelectron spectroscopy (XPS)

## Future

* Docmuentation (priority)
* Add database of inelastic mean free path (IMFP) of common materials

## X-ray Photonelectron Spectroscopy

### Example

An example analyzing a Ge peak fit within CasaXPS.

    import xps

    # Shortcut for the XPS machine at Dalhousie
    # xps_mach = 'Dalhousie'
    # mach = xps.MACH_PARAM[xps_mach]
    mach = {
        'coef' : [0.9033, -4.0724, 5.0677, 1.1066],
        'scale'     : 0.01,
        'work_func' : 4.6,
    }
    hv = xps.PHOTON_ENERGY['Al']

    # Pass energy [eV], found from XPS operator
    pe = 30

    sfwagner_Ge = xps.sf.Ge['3d']['area']

    data = xps.parser.CasaXPS('Ge_example.csv')

    # Labels defined by user in CasaXPS fits
    pk_lbl = 'Ge 3d'
    be = data.binding_energy(pk_lbl)
    area = data.area(pk_lbl)

    ke = xps.kinetic_energy(be, hv, mach['work_func'])

    t_fn = xps.transmission(ke, pe, mach['coef'], mach['scale'])
    t_fn_wagner = 1/ke # Proportional to 1/KE

    sf_mach = xps.sf_machine(sfwagner_Ge, t_fn, t_fn_wagner)

    pk_corr = xps.peak_correction(area, sf_mach)

    # NOTE Use the xps.XPSPeak(...) helper for convienence
    # analyzed_Ge = xps.XPSPeak(pk_lbl, be, area, sfwagner_Ge, hv, pe, mach)

    # NOTE Returns pandas.DataFrame with all parameters calculated.
    # The user can also query parameters individually
    # df = analyzed_Ge.get_data()

### Matrix Factor corrections

If using multple elements within a matrix (e.g. an alloy), you can utilize the
`xps.matrix_factor` function. You require the inelastic mean free path
of electron scattering (*imfp*) of both species in bulk and the density, as
well as the *imfp* of the matrix at the measured kinetic energies of both
elements. For example, if you have two corrected peaks: `pk_Mn_corr`, and
`pk_Ge_corr`. The *imfp* can be calculated using the TPP-2M equation for
inelastic mean free path, found in the following reference:

S. Tanuma, C. J. Powel, D. R. Penn, *Surf. Interf. Anal.*, Vol 21, 165 (1994)


    import xps
    # pk_Mn_corr and pk_Ge_corr calculated as in the example above

    # a is the kinetic energy used to determine imfp of Ge in Bulk
    imfp_matrix_a = 21.17
    imfp_Ge_a = 29.84
    rho_Ge_a = 5.32

    # b is the kinetic energy used to determine imfp of Mn in Bulk
    imfp_matrix_b = 14.17
    imfp_Mn_b = 14.87
    rho_Mn_b = 7.43

    mat_fact = xps.matrix_factor(imfp_Ge_a, imfp_Mn_b,
                                 mfp_matrix_a, mfp_matrix_b,
                                 rho_Ge_a, rho_Mn_b)
    relative_pk_Ge_corr = (pk_Ge_corr/pk_Mn_corr)*mat_fact

    # NOTE because Mn is used as the normalizing component we can use its
    # corrected peak value directly, all other elements require the matrix
    # factor correction
    print('Ratios of Mn and Ge in MnGe matrix')
    print('Mn : {:0.4e}'.format(pk_Mn_corr))
    print('Ge : {:0.4e}'.format(relative_pk_Ge_corr))

### sfwagner.{db,py}: Empirically derived set of atomic sensitivity factors for XPS

The data in Appendix 5 is reproduced and provided here for non-profit use with
permission of the publisher John Wiley & Sons Ltd.

"Practical Surface Analysis by Auger and X-ray Photoelectron Spectroscopy",
D. Briggs and M. P. Seah,
Appendix 5, p511-514,
Published by J. Wiley and Sons in 1983, ISBN 0-471-26279

Copyright (c) 1983 by John Wiley & Sons Ltd.

The original set of data first appeared in the following resource:
C. D. Wagner, L. E. Davis, M. V. Zeller, J. A. Taylor, R. M. Raymond and L. H. Gale,
Surf. Interface Anal., 3. 211 (1981)

Any use of this data must include the citations above in any work.

## Electron Inelastic Mean Free Path (IMFP)
Electron IMFP can be calculated from using the Tanuma, Powel, Penn modified
(TPP-2M) equation derived from equations (3), (4b,c,d,e) and (8) in the
following reference:

S. Tanuma, C. J. Powel, D. R. Penn, *Surf. Interf. Anal.*, Vol 21, 165 (1994)

For convienence the IMFP TPP-2M equation is located in `xps.scatter` and
can be used as such:

    from xps import scatter

    # Mn example
    kinetic_energy = 1000 # Can be calculated from xps.kinetic_energy

    rho = 7.43         # [g/cc]
    Nv = 7             # valence electrons
    M = 53.938         # atomic mass
    bandgap_energy = 0 # [eV]

    # Return SI units [m]
    imfp_Mn = scatter.imfp_TPP2M(kinetic_energy, rho, M, Nv,
                                 bandgap_energy, 'SI')

The value here can be used in the `xps.matrix_factor` calculations
outlined above.


CHANGELOG.md
# Change Log

## v0.2.2
    * Switched to pathlib.Path
    * Code cleanup via flake8
    * Moved to hatch build system for python PEP >=3.10 compliance
    * Cleaned up ambiguity in math

## v0.2.1
    * Added doc strings
    * Made pythonic changes

## v0.2.0

### Changed
    * Renamed module from surfsci -> xps with it's own repositories.
      The reason for the change is to keep the scope focused on XPS.
    * Syntax change with sensitivity factors for wagner xps.sf.<ELEM>

## v0.1.7

### Added
    * Added IMFP TPP-2M functionality, lacks database of materials
    * Initial usage documentation

## v0.1.6

### Added
    * Added Wagner sensitivity factors with copyright attribution

## v0.1.5

### Added
    * xps.py: added the ability to deal with matrix of material
    * Added column parsing of CasaXPS data files

## v0.1.4

### Fixed
    * xps.XPSPeak minor syntax fixes

## v0.1.3

### Added
    * Basic parser for CasaXPS data files

## v0.1.2

### Added
    * Helper class xps.XPSPeak added for ease of analysis

## v0.1.0

### Added
    * Basic usage available


src/xps/xps.py
# ISC License
#
# Copyright 2018 David Kalliecharan <dave@dal.ca>
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
# SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
# OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
# CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
"""
xps.py
------

X-Ray photoelectron spectroscopy (XPS) helper functions to adjust the
machine dependent intensities to Wagner intensities
"""

from numpy import log10
from numpy import poly1d
from pandas import DataFrame

__author__  = 'David Kalliecharan'
__license__ = 'ISC License'
__status__  = 'Development'

PHOTON_ENERGY = {
    'Mg' : 1253.6,  # eV
    'Al' : 1486.6,  # eV
}

MACH_PARAM = {
    'Dalhousie': {
        'coef'      : [0.9033, -4.0724, 5.0677, 1.1066],
        'scale'     : 0.01,
        'work_func' : 4.6,
    }
}


def kinetic_energy(be, hv, psi):
    """Calculates the kinetic energy (KE) with a Mg source (Dalhousie default):

    be  : binding energy (BE [eV])
    psi : Work function (Psi [eV])
    hv  : Phonton energy (h*nu [eV])
    """
    return hv - be - psi


def transmission(ke, pe, a, scale):
    """Dalhousie standard for the Transmission function

    ke    : Kinetic energy (KE [eV])
    pe    : Pass Energy (PE [eV])
    a     : List of coefficients for series expansion, _NOTE_ coefficents are
            reversed such that (aN, ..., a0) -> aN * x ** N + ... + a0
            See documentation on numpy.poly1d
    scale : required scaling factor according to Avantage software
    """
    f = poly1d(a)
    x = log10(ke / pe)
    y = f(x)
    return scale * pe * 10 ** y


def sf_machine(sf_wagner, transmission_mach, transmission_wagner):
    """Sensitivity factor for XPS machine correction from the Wagner XPS machine

    sf_wagner              : Wagner sensitivity factor
    transmission_mach      : Transmission function for the selected element
    transmission_wagner    : Wagner transmission function for the selected
                             element.
                             Proportional to 1/(kinetic_energy) [1/eV]
    """
    return sf_wagner * transmission_mach / transmission_wagner


def peak_correction(peak_area, sf_mach):
    """Corrected intensity of peak fitting by scaling to the sensitivity
    factor of the machine.

    peak_area   : peak area [CPS*eV]
    sf_mach     : sensitivity factor of xps machine
    """
    return peak_area / sf_mach


def matrix_factor(mfp_a, mfp_b, mfp_matrix_a, mfp_matrix_b, rho_a, rho_b):
    """Determines the matrix multiplicity factor. Mean free path values can be
found using an external program QUASES-IMFP-TPP2M.

    mfp_a        : Mean free path (mfp) of a in bulk, eg. mfp_a(KE_{Mn 2p3/2})
    mfp_b        : Mean free path (mfp) of b in bulk, eg. mfp_b(KE_{Co 2p3/2})
    mfp_matrix_a : Mean free path of matrix with kinetic energy of a
                   mfp_matirx(KE_{Mn 2p3/2})
    mfp_matrix_b : Mean free path of matrix with kinetic energy of b
                   mfp_matirx(KE_{Co 2p3/2})
    rho_a        : Density of species a
    rho_b        : Density of species b
    """
    a = mfp_matrix_a / mfp_a / rho_a
    b = mfp_matrix_b / mfp_b / rho_b
    return a / b


class XPSPeak():
    """XPSPeakBase class that conveniently wraps the base functions into an object
    peak_id    : Peak ID label
    be         : Binding energy [eV]
    peak_area  : Area of peak [CPS * eV]
    sf_wag     : Wagner sensitivity factor
    hv         : Photon energy
    pe         : Pass energy
    mach_param : Dictionary of XPS machine parameters:
                 coef      : [aN, aN-1, ..., a1, a0]
                 work_func : float [eV]
                 scale     : float
    """
    def __init__(self, peak_id, be, peak_area, sf_wagner, hv, pe, mach_param,
                 *args, **kws):
        self.__peak_id = peak_id
        self.__be = be
        self.__peak_area = peak_area
        self.__sf_wagner = sf_wagner
        self.__hv = hv
        self.__pe = pe
        self.__a = mach_param['coef']
        self.__psi = mach_param['work_func']
        self.__scale = mach_param['scale']
        self.__args = args
        self.__kws = kws

    def get_kinetic_energy(self):
        """Get kinetic energy at a binding energy, see xps.kinetic_energy"""
        return kinetic_energy(self.__be, self.__hv, self.__psi)

    def get_transmission(self):
        """Get transmission function, see xps.transmission"""
        ke = self.get_kinetic_energy()
        return transmission(ke, self.__pe, self.__a, self.__scale)

    def get_sf_machine(self):
        """Get sensitivity factor of machine, see xps.sf_machine"""
        ke = self.get_kinetic_energy()
        transmission_wagner = 1 / ke  # proportional to 1/KE
        if 'transmission_wagner' in self.__kws:
            self.transmission_wagner = self.__kws['transmission_wagner']

        tx_xps = self.get_transmission()
        return sf_machine(self.__sf_wagner, tx_xps, transmission_wagner)

    def get_peak_correction(self):
        """Get peak corrections, see xps.peak_correction"""
        sf_mach = self.get_sf_machine()
        return peak_correction(self.__peak_area, sf_mach)

    def get_mach_params(self):
        """Get machine paramters"""
        params = {
            'photon_energy' : self.__hv,
            'pass_energy'   : self.__pe,
            'coefficients'  : self.__a,
            'scale'         : self.__scale,
            'work_func'     : self.__psi,
        }
        return params

    def set_matrix_component(self, mfp, mfp_matrix, rho):
        """
    mfp        : Mean free path (mfp) in bulk, eg. mfp_a(KE_{Mn 2p3/2})
    mfp_matrix : Mean free path of matrix with kinetic energy of a
                   mfp_matirx(KE_{Mn 2p3/2})
    rho        : Density of material
        """
        self.__matrix_component = mfp_matrix / mfp / rho

    def get_data(self, return_dict=False):
        data_dict = {
            'peak_id'           : [self.__peak_id],
            'binding_energy'    : [self.__be],
            'kinetic_energy'    : [self.get_kinetic_energy()],
            'transmission_func' : [self.get_transmission()],
            'sf_wagner'         : [self.__sf_wagner],
            'sf_machine'        : [self.get_sf_machine()],
            'peak'              : [self.__peak_area],
            'peak_corrected'    : [self.get_peak_correction()],
        }

        if hasattr(self, '_XPSPeak__matrix_component'):
            data_dict['matrix_component'] = [self.__matrix_component]

        if return_dict is True:
            return data_dict
        return DataFrame(data=data_dict)


if __name__ == '__main__':
    pass
